src/interaction/FixedLocalTupleComListInteractionTemplate.hpp

```diff
@@ -0,0 +1,408 @@
+		for (FixedLocalTupleList::TupleList::Iterator it(*fixedtupleList); it.isValid(); ++it) {
+		    Particle *p = it->first;
+		    pid_list[pidK] = p->id();
+		    
+		    for(long unsigned int i = 0; i < pidK; i++) {
+			if (pid_list[i]/N_Constrain == pid_list[pidK]/N_Constrain) {
+			    std::stringstream msg;
+			    msg << "ERROR: Particle ID is redundant\n";
+			    err.setException(msg.str());
+			}
+		    }
+		    pidK++;
+		}
+
+		Real3D* totCOM;
+		totCOM = new Real3D[num_of_subchain];
+		Real3D* COM;
+		COM = new Real3D[num_of_subchain];
+		
+		Real3D Li = system.bc->getBoxL();
+		
+		for(long unsigned int i = 0; i < num_of_subchain; i++) {
+		    COM[i] = 0.;
+		    totCOM[i] = 0.;
+		}
+
+		for (FixedLocalTupleList::TupleList::Iterator it(*fixedtupleList); it.isValid(); ++it) {
+		    Particle *p = it->first;
+		    Real3D pos_i = p->position();
+		    Real3D tmp_com = pos_i;
+		    std::vector<Particle*> pList = it->second;
+
+		    for (long unsigned int j = 1; j < N_Constrain; j++) {
+			p = pList[j - 1];
+			Real3D pos_j = p->position();
+			Real3D dist;
+			system.bc->getMinimumImageVectorBox(dist, pos_j, pos_i);
+			pos_j = pos_i + dist;
+			tmp_com += pos_j;
+			pos_i = pos_j;
+		    }
+		    COM[(p->id() - 1)/N_Constrain] = tmp_com/N_Constrain;
+		}
+
+		boost::mpi::all_reduce(*system.comm, COM, num_of_subchain, totCOM, std::plus<Real3D>() );
+		
+		for (long unsigned int i = 0; i < num_of_subchain; i++) {
+		    com_origin[i] = totCOM[i];
+		}
+		
+		delete [] COM;
+		COM = NULL;
+		delete [] totCOM;
+		totCOM = NULL;
+	    }
+	    
+	    virtual ~FixedLocalTupleComListInteractionTemplate() {};
+	    
+	    // set the center of mass of subchains
+	    void setCom(longint id, const Real3D& pos) {
+		com_origin[id] = pos;
+	    }
+
+	    void
+	    setFixedLocalTupleList(shared_ptr < FixedLocalTupleList > _fixedtupleList) {
+		fixedtupleList = _fixedtupleList;
+	    }
+	    
+	    shared_ptr < FixedLocalTupleList > getFixedLocalTupleList() {
+		return fixedtupleList;
+	    }
+	    
+	    void
+	    setPotential(shared_ptr < Potential> _potential) {
+		if (_potential) {
+		    potential = _potential;
+		} else {
+		    LOG4ESPP_ERROR(theLogger, "NULL potential");
+		}
+	    }
+	    
+	    shared_ptr < Potential > getPotential() {
+		return potential;
+	    }
+	    
+	    virtual void addForces();
+	    virtual real computeEnergy();
+	    virtual real computeEnergyDeriv();
+	    virtual real computeEnergyAA();
+	    virtual real computeEnergyCG();      
+	    virtual void computeVirialX(std::vector<real> &p_xx_total, int bins); 
+	    virtual real computeVirial();
+	    virtual void computeVirialTensor(Tensor& w);
+	    virtual void computeVirialTensor(Tensor& w, real z);
+	    virtual void computeVirialTensor(Tensor *w, int n);
+	    virtual real getMaxCutoff() { return 0.0; }
+	    virtual int bondType() { return Nonbonded; }
+	    
+	protected:
+	    std::map<long unsigned int, Real3D> computeCOM() {
```

Comment:
Yes. This function calculate COM of subchains and I assumed that the mass of all particles is one.
The equation of calculating COM of subchain $\vec{r_{com}}$ is defined as:
$\vec{r_{com}} = \frac{1}{N} \sum^N_{i=1} \vec{r_{i}}$.
In the lines 206:211, $\vec{r_{i+1}}$ is described as $\vec{r_i} + (\vec{r_{i+1}} - \vec{r_i})$.
The variable `dist` stands for just $\vec{r_{i+1}} - \vec{r_i}$.
When we use open boundary, this expression is nonsense.
In contrast, under periodic boundary condition, this expression would be meaningful. Since, the influence of periodic boundary condition can be removed at calculating $\vec{r_{i+1}} - \vec{r_i}$.
Does that make sense?